### PR TITLE
systemd integration with container runtime for supporting sd_notify protocol

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan shishir.mahajan@redhat.com

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
ping @mrunalp

**Comment by mrunalp** [comment]
@shishir-a412ed Could you add some description about how this works and is useful? Thanks!

**Comment by rhatdan** [comment]
@shishir-a412ed Add a better description to your pull request of what sd_notify does and why this is important.

sd_notify stands for systemd notify, it basically allows a containerized service to tell systemd that it is fully up and ready.

An example of this is the cups daemon., you might want to setup services that rely on cups printing, but you do not want them to send print jobs until cups is fully up and running.   You could setup cups as a notify unit file, and then other services that rely on cups, will not be started by systemd until the process within the cups container writes to the sd_notify service.  

Currently you can not do this with docker clients.  BTW Docker daemon itself does sd_notify.

File: run.go [hunk 1]
```diff
@@ -67,6 +67,12 @@ func execContainer(context *cli.Context, spec *specs.LinuxSpec) (int, error) {
 // default action is to execute a container
 func runAction(context *cli.Context) {
 	spec, err := loadSpec(context.Args().First())
+
+	if os.Getenv("NOTIFY_SOCKET") != "" {
```
Comment (mrunalp) [inline run.go]: Could you store the value in a variable so we make the os.Getenv call only once instead of 4 times?

Comment (mrunalp) [inline run.go]: Also, this logic could be moved to a separate function.

**Comment by shishir-a412ed** [comment]
@mrunalp sorry I should have added an elaborate description.
Here it goes !

systemd notify a.k.a sd_notify

Problem:
Currently if a service is containerized and is using sd_notify protocol. 
There is no way for the service to notify systemd of its service startup completion.
Not unless we have systemd capabilities within the container or a way to communicate from within the container to the systemd on host.

Use-case:
systemd needs to start two services.
a) a backend database like mongoDb, cassandra within a container.
b) an apache front end web server on the host.

systemd should not start apache webserver until the backend database is
up and running.

Solution:
systemd uses NOTIFY_SOCKET environment variable for sd_notify protocol.

1) if NOTIFY_SOCKET is set on host
     a) bind mount the socket file on host into the container.
     b) Leak the environment variable NOTIFY_SOCKET into the container.

once the backend database is completely up and running. it will make the sd_notify() call.
Only this time, the socket is available in the container.
Systemd will be listening on that socket and will get a message (READY=1).

Now systemd can start the apache webserver.

Without this patch systemd will never get back the READY=1, response so
it will never start the apache webserver.

cups is an example of a systemd sd_notify service. Even the docker
daemon supports the sd_notify protocol.

Here is the list of services that are using this protocol on my Fedora
system.

./cups.service:Type=notify
./systemd-networkd.service:Type=notify
./systemd-nspawn@.service:Type=notify
./systemd-udevd.service:Type=notify
./smb.service:Type=notify
./systemd-resolved.service:Type=notify
./docker.service:Type=notify
./libvirtd.service:Type=notify
./nmb.service:Type=notify
./docker.service.old:Type=notify
./user@.service:Type=notify
./systemd-timesyncd.service:Type=notify
./httpd.service:Type=notify
./systemd-journald.service:Type=notify

**Comment by rhatdan** [comment]
@shishir-a412ed Add this comment to your git commit, so people would have a record, I also agree with @mrunalp on using a local variable and moving this to a function.

**Comment by mrunalp** [comment]
@LK4D4 @crosbymichael PTAL. I think this provides good integration with systemd with minimal code.

**Comment by marcosnils** [comment]
@shishir-a412ed good catch, this is clean and really useful!.

**Comment by crosbymichael** [comment]
LGTM

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**